qiskit/visualization/circuit/_utils.py

```diff
@@ -397,8 +401,15 @@ def _get_layered_instructions(
     if justify:
         justify = justify.lower()
 
-    # default to left
-    justify = justify if justify in ("right", "none") else "left"
+    # If wrong input, default ('left') will be used.
+    if justify not in ("left", "right", "none"):
+        if justify is not None:
+            warn(
+                f"justify must be 'left, right or none', not: '{justify}'. Default (left) will be used.",
+                UserWarning,
+                2,
+            )
```

Comment:
@1ucian0 @mtreinish can we error here instead of warn? I'm not sure if the deprecation policy allows it: https://github.com/Qiskit/qiskit/blob/main/DEPRECATION.md#changing-behavior

This code path was always broken of using an unrecognized arg for `justify`. Before, we silently switched to the fallback. Can we now error?

Reply:
Whatever you guys prefer/decide is okeywith me! But I guess raising an `error` is less backwards compatible, at least for people with bad `justify`s around there, that didn't know we were secretly changing it 😅.

Also notice, that the corner case I mentioned in the description, of manually inputting `justify=None`, will still not be covered in either case, since it's the default value all the way up... and we can't `warn`/`raise error` for it.

Personally I would change the default to the real one: `left` in cascade up, but I would like to know your opinions first, since that would be a bigger change.

Reply:
That's a good point about `justify=None`. I like your suggestion to change the default to be `justify="left"`. That is more obvious code to understand, including generating more clear documentation at https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.QuantumCircuit#draw. It also allows us to warn/error when explicitly setting `=None`, like you suggest. 

For now, please still keep the explicit fallback to `left` (line 412). But you could remove `if justify is not None:` because we can now always warn if the value is not already one of the 3 valid options.

You'll also want to update all the functions that (transitively) call `_get_layered_instructions` to use the new default of `="left"`.

--

After thinking about it more, I agree with you that it would not yet be safe to error because it would result in someone's program that used to work no longer running. 

However, erroring is still the better behavior in general. So, we can introduce a deprecation so that we can switch to erroring in 2.0.

The best way to deprecate this is to use https://docs.quantum.ibm.com/api/qiskit/utils#deprecate_arg. You'll use that on every _public_ interface where a user can set `justify`, like `QuantumCircuit.draw`. It's useful to put the deprecations closer to where the users set the argument so that you can make the message more clear of where the bad code is coming from. You don't want to set it on all the private helper functions because it would be noisy to have the message in both places; so you would actually remove the `warning` inside `_get_layered_instructions`. 

You'll want to set `deprecation_description` to something like `Setting QuantumCircuit.draw()’s justify argument to a value other than 'left', 'right', or 'none'`. You also need to set `predicate`.

Reply:
Great! 
Those changes were exactly what I was thinking 👌, I'll apply them when I have a free moment 👍

--

And respect the deprecation, yeah I think that is definetly the way to go, honestly didn't know how you did it, so I didn't want to jump into deprecations, but with your instructions and that wrapper, this will be much easier, thanks!

Reply:
### I have a question, regarding the `@deprecate_arg`: 

In this concrete case, we have `Circuit.draw()` and `circuit_drawer()`, where the first, calls the second one, but both are public... 

--- 
So if I add the wrapper in both public methods, I will get the error twice for `Circuit.draw()`, like in 🟨 here:
![Screenshot 2024-06-06 at 02 20 01](https://github.com/Qiskit/qiskit/assets/109400222/d836a004-5751-42e4-bdce-cf9d005d483f)

But if instead, I only leave the wrapper in the common one (`circuit_drawer()`), I won't get two warns, but I won't get a very accurate position of the error, like in 🟨 here:
![Screenshot 2024-06-06 at 02 19 41](https://github.com/Qiskit/qiskit/assets/109400222/56f5de04-ef4b-4f8c-af95-3bdecec41116)

### I'm not sure what is more desirable, I'll leave the first one until I receive further feedback 👍

Reply:
This is a little convoluted, but you can set the default value to `left` when `QuantumCircuit.draw()` calls `circuit_drawer`. So, you fix the bad arg before the second warning is relevant.

You'll be repeating this logic multiple times to determine if `justify` is an invalid value. So you can define a private helper function like `_is_invalid_justify_arg`.

Reply:
Ohh, I see, that makes sense, great! 👍

Reply:
Did it!

The only problem I see with my implementation is the repetitive use of the `.lower()` method, but I hope that is acceptable, since the other ways I tried don't look as readable...

---

Also, mention that this solves a previous error, for when the user didn't pass a `string` neither `None`, for which `justify.lower()`, would have raised the standard non-personalized error.